core/unit_test/incremental/Test05_MDRangePolicy.hpp

```diff
@@ -0,0 +1,197 @@
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:5
+// Unit Test for MDRangePolicy without Views uptil 4 ranks.
+// For each of the MDRangePolicy test from 2-to-4 ranks, we create an equivalent
+// dimensional view. In each of these views we update the
+// elements as a product of iterator indexes and a constant inside a
+// parallel_for lambda. At the end, we check for correctness.
+
+namespace Test {
+
+using value_type       = double;
+const int N            = 10;
+const int M            = 10;
+const value_type delta = 0.5;
+
+template <class ExecSpace>
+struct TestMDRangePolicy {
+  // 2D View
+  using View_2D      = typename Kokkos::View<value_type **, ExecSpace>;
+  using Host_View_2D = typename View_2D::HostMirror;
+  Host_View_2D hostDataView_2D;
+
+  // 3D View
+  using View_3D      = typename Kokkos::View<value_type ***, ExecSpace>;
+  using Host_View_3D = typename View_3D::HostMirror;
+  Host_View_3D hostDataView_3D;
+
+  // 4D View
+  using View_4D      = typename Kokkos::View<value_type ****, ExecSpace>;
+  using Host_View_4D = typename View_4D::HostMirror;
+  Host_View_4D hostDataView_4D;
+
+  // Memory space type for Device and Host data
+  using d_memspace_type = typename ExecSpace::memory_space;
+  using h_memspace_type = Kokkos::HostSpace;
+
+  // Index Type for the iterator
+  using int_index = Kokkos::IndexType<int>;
+
+  // An MDRangePolicy for 2 nested loops
+  using MDPolicyType_2D = typename Kokkos::Experimental::MDRangePolicy<
+      ExecSpace, Kokkos::Experimental::Rank<2>, int_index>;
+
+  // An MDRangePolicy for 3 nested loops
+  using MDPolicyType_3D = typename Kokkos::Experimental::MDRangePolicy<
+      ExecSpace, Kokkos::Experimental::Rank<3>, int_index>;
+
+  // An MDRangePolicy for 4 nested loops
+  using MDPolicyType_4D = typename Kokkos::Experimental::MDRangePolicy<
+      ExecSpace, Kokkos::Experimental::Rank<4>, int_index>;
+
+  // compare and equal
+  void compare_equal_2D() {
+    for (int i = 0; i < N; ++i)
+      for (int j = 0; j < M; ++j)
+        ASSERT_EQ(hostDataView_2D(i, j), i * j * delta);
```

Comment:
lets fill this with something which is unique per pair `(i*M+j)` for example

Reply:
Well i*j*delta is unique per (i,j) pair.

Reply:
Christian is pointing out that `i * j * delta` equals `j * i * delta`, so you can't tell the difference between the View and its transpose.

Reply:
i * N + j might be what we want here. Let's scratch the scale factor (delta)

Reply:
jup exactly.

Reply:
Ah yes, ok will update that.

Reply:
Retest this please.